node/src/components/rpc_server/rpcs/state.rs

```diff
@@ -558,3 +576,262 @@ impl RpcWithParamsExt for GetAccountInfo {
+        &*GET_DICTIONARY_ITEM_PARAMS
+    }
+}
+
+/// Result for "state_get_dictionary" RPC response.
+#[derive(Serialize, Deserialize, Debug, JsonSchema)]
+#[serde(deny_unknown_fields)]
+pub struct GetDictionaryItemResult {
+    /// The RPC API version.
+    #[schemars(with = "String")]
+    pub api_version: ProtocolVersion,
+    /// The key under which the value is stored.
+    pub dictionary_key: String,
+    /// The stored value.
+    pub stored_value: StoredValue,
+    /// The merkle proof.
+    pub merkle_proof: String,
+}
+
+impl DocExample for GetDictionaryItemResult {
+    fn doc_example() -> &'static Self {
+        &*GET_DICTIONARY_ITEM_RESULT
+    }
+}
+
+/// "state_get_dictionary" RPC.
+pub struct GetDictionaryItem {}
+
+impl RpcWithParams for GetDictionaryItem {
+    const METHOD: &'static str = "state_get_dictionary_item";
+    type RequestParams = GetDictionaryItemParams;
+    type ResponseResult = GetDictionaryItemResult;
+}
+
+impl RpcWithParamsExt for GetDictionaryItem {
+    fn handle_request<REv: ReactorEventT>(
+        effect_builder: EffectBuilder<REv>,
+        response_builder: Builder,
+        params: Self::RequestParams,
+        api_version: ProtocolVersion,
+    ) -> BoxFuture<'static, Result<Response<Body>, Error>> {
+        async move {
+            let dictionary_address = match params.dictionary_identifier {
+                DictionaryIdentifier::AccountNamedKey { .. }
+                | DictionaryIdentifier::ContractNamedKey { .. } => {
+                    let base_key = match params.dictionary_identifier.get_dictionary_base_key() {
+                        Ok(Some(key)) => key,
+                        Err(_) | Ok(None) => {
+                            return Ok(response_builder.error(warp_json_rpc::Error::custom(
+                                ErrorCode::ParseQueryKey as i64,
+                                "Failed to parse key",
+                            ))?)
+                        }
+                    };
+
+                    let path = match params.dictionary_identifier.get_base_query_path() {
+                        Ok(Some(path)) => path,
+                        Err(_) | Ok(None) => {
+                            return Ok(response_builder.error(warp_json_rpc::Error::custom(
+                                ErrorCode::NoDictionaryName as i64,
+                                "Failed to execute query",
+                            ))?)
+                        }
+                    };
+
+                    let query_result = effect_builder
+                        .make_request(
+                            |responder| RpcRequest::QueryGlobalState {
+                                state_root_hash: params.state_root_hash,
+                                base_key,
+                                path,
+                                responder,
+                            },
+                            QueueKind::Api,
+                        )
+                        .await;
```

Comment:
We shoulde use `EffectBuilder::query_global_state` here

Reply:
I feel like keeping it as is, `EffectBuilder::query_global_state` requires building a `QueryRequest` and this way the request is it gets built for me, so I'd prefer to keep it this way

Reply:
Karan actually did this "correctly", per the original design of the rpc component. This causes the RpcServer component to deal with it in it's handle_event  vs the http filter hook. Kind of silly IMO, but that is how it was meant to work afaik.